sam_handler/handler.py

```diff
@@ -0,0 +1,247 @@
+        credentials: Credentials = self.connection_manager._get_session(
+            self.connection_manager.profile,
+            self.connection_manager.region,
+            self.connection_manager.iam_role
+        ).get_credentials()
+        envs.update(
+            AWS_ACCESS_KEY_ID=credentials.access_key,
+            AWS_SECRET_ACCESS_KEY=credentials.secret_key,
+        )
+
+        # There might not be a session token, so if there isn't one, make sure it doesn't exist in
+        # the envs being passed to the subprocess
+        if credentials.token is None:
+            envs.pop('AWS_SESSION_TOKEN', None)
+        else:
+            envs['AWS_SESSION_TOKEN'] = credentials.token
+
+        return envs
+
+
+class SAM(TemplateHandler):
+    """A template handler for AWS SAM templates. Using this will allow Sceptre to work with SAM to
+    build and package a SAM template and deploy it with Sceptre.
+    """
+
+    SAM_ARTIFACT_DIRECTORY = 'sam_artifacts'
+
+    def __init__(
+        self,
+        name,
+        arguments=None,
+        sceptre_user_data=None,
+        connection_manager=None,
+        stack_group_config=None,
+        *,
+        invoker_class=SamInvoker,
+        get_temp_dir=tempfile.gettempdir
+    ):
+        super().__init__(name, arguments, sceptre_user_data, connection_manager, stack_group_config)
+        self.invoker_class = invoker_class
+        self.get_temp_dir = get_temp_dir
+
+    def schema(self) -> dict:
+        """This is the json schema of the template handler. It is required by Sceptre to define
+        template handler parameters.
+        """
+        return {
+            "type": "object",
+            "properties": {
+                "sam_template_path": {"type": "string"},
+                "artifact_prefix": {"type": "string"},
+                "artifact_bucket_name": {"type": "string"},
+                "sam_build_args": {"type": "object"},
+                "sam_package_args": {"type": "object"}
+            },
+            "required": [
+                "sam_template_path",
+                "artifact_bucket_name",
+            ]
+        }
+
+    def handle(self) -> str:
+        invoker = self.invoker_class(
+            connection_manager=self.connection_manager,
+            sam_directory=self.sam_directory
+        )
+        self._create_generation_destination()
+        self._build(invoker)
+        self._package(invoker)
+        return self.destination_template_path.read_text()
+
+    @property
+    def sam_template_path(self) -> Path:
+        return Path(self.arguments['sam_template_path']).absolute()
+
+    @property
+    def sam_directory(self) -> Path:
+        return self.sam_template_path.parent
+
+    @property
+    def destination_template_path(self) -> Path:
+        suffix = self.sam_template_path.suffix
+        path_segments = self.name.split('/')
+        path_segments[-1] += suffix
+        return Path(self.get_temp_dir()).joinpath(*path_segments).absolute()
```

Comment:
By default the SAM CLI creates builds in a `.aws-sam` folder.  why not just keep that as the default?

Reply:
I chose not to rely on that because there are various ways to configure sam to build to a different location. Also, I'd rather be specific about where I want the file than rely on default behavior.